### DISCUSSION TITLE
Building a Complete Toolchain

### DISCUSSION THREAD

**Discussion by thekief** [android/ndk#1890] [Q&A] [answered]
I wanted to introduce some changes to the LLVM toolchain for testing purposes and built a toolchain following the official guide [1]. I stumbled on an issue though, the wrapper bash scripts for clang, e.g.,

```
./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi22-clang
./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/aarch64-linux-android30-clang
./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/aarch64-linux-android27-clang
./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi27-clang
./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/x86_64-linux-android21-clang
./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/i686-linux-android24-clang
./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/i686-linux-android16-clang
./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/i686-linux-android17-clang
./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi28-clang
./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/x86_64-linux-android27-clang
./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/i686-linux-android29-clang
./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi17-clang
./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/i686-linux-android23-clang
./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/i686-linux-android18-clang
./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/x86_64-linux-android31-clang
./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/x86_64-linux-android22-clang
./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/i686-linux-android22-clang
./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/x86_64-linux-android23-clang
./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/i686-linux-android21-clang
./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi23-clang
./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/aarch64-linux-android24-clang
```

are not created. I looked into this and noticed that `toolchain/prebuilts/ndk/r24/build/tools/make_standalone_toolchain.py` builds thems. Further down it is written that:

`make_standalone_toolchain.py is no longer necessary`

So I was wondering if there is a new way to create the bash script wrapper or `make_standalone_toolchain.py` is still the way to go.

[1] https://android.googlesource.com/toolchain/llvm_android/+/master/README.md

**Comment by rprichard** [comment, marked as answer]
Running build.py in llvm_android creates a "toolchain prebuilt", like the `clang-r487747c` directory in:
https://android.googlesource.com/platform/prebuilts/clang/host/linux-x86/+/refs/heads/master

The wrapper scripts with the API levels in them (the `24` in `aarch64-linux-android24-clang`) are actually created by the NDK build script ([checkbuild.py](https://android.googlesource.com/platform/ndk/+/refs/tags/ndk-r25c/ndk/checkbuild.py#1445), https://android.googlesource.com/platform/ndk/+/master/docs/Building.md).

See the "NDK" section in https://android.googlesource.com/platform/prebuilts/clang/host/linux-x86/+/master/README.md, which describes how to update the NDK build script to use a different toolchain prebuilt.

The make_standalone_toolchain.py script copies the NDK and creates a few extra wrapper scripts like `clang` and `clang++` scripts that target a specific Android triple. It's been more-or-less unnecessary for a few years now, because the triple-prefixed wrapper scripts created by checkbuild.py should be sufficient.

**Comment by nikijohnson77** [comment]
> I wanted to introduce some changes to the LLVM toolchain for testing purposes and built a toolchain following the official guide [1]. I stumbled on an issue though, the wrapper bash scripts for clang, e.g.,
> 
> ```
> ./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi22-clang
> ./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/aarch64-linux-android30-clang
> ./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/aarch64-linux-android27-clang
> ./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi27-clang
> ./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/x86_64-linux-android21-clang
> ./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/i686-linux-android24-clang
> ./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/i686-linux-android16-clang
> ./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/i686-linux-android17-clang
> ./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi28-clang
> ./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/x86_64-linux-android27-clang
> ./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/i686-linux-android29-clang
> ./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi17-clang
> ./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/i686-linux-android23-clang
> ./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/i686-linux-android18-clang
> ./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/x86_64-linux-android31-clang
> ./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/x86_64-linux-android22-clang
> ./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/i686-linux-android22-clang
> ./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/x86_64-linux-android23-clang
> ./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/i686-linux-android21-clang
> ./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi23-clang
> ./ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/aarch64-linux-android24-clang
> ```
> 
> are not created. I looked into this and noticed that `toolchain/prebuilts/ndk/r24/build/tools/make_standalone_toolchain.py` builds thems. Further down it is written that:
> 
> `make_standalone_toolchain.py is no longer necessary`
> 
> So I was wondering if there is a new way to create the bash script wrapper or `make_standalone_toolchain.py` is still the way to go.
> 
> 
> [1]__ https://android.googlesource.com/toolchain/llvm_android/+/master/REAjDME.md
